Fixes #6153 Add case typechecking to provided props against propTypes

brad-decker opened the pull request:
Grabs mismatched values compared to PropTypes, can be used later to warn against props being provided that aren't defined in PropTypes.

It then uses toLowerCase to find similar props and warn if there is a defined propType with the same composition but difference case as a prop passed on the element. 

Warning follows the "did you mean %s?" type suggestions in other react warnings. 
# reactEurope-hackathon

facebook-github-bot-0 commented:
@brad-decker updated the pull request.

gaearon commented:
This seems to achieve the same goal as https://github.com/facebook/react/pull/6255. It doesn’t seem to dedupe warnings though. In general, I’m not sure if we want this. See https://github.com/facebook/react/pull/6255#issuecomment-228619578 for some comments. Let’s first reach a decision on that, and then work on PRs.

brad-decker commented:
@gaearon Thanks for the response. I'm fine with waiting for a decision on this and closing this PR if there is a better solution or one more inline with the direction you want to go.

gaearon commented:
Can you please make this deduplicate warnings? I would probably go with this rather than with a full-blown devtool approach in this particular case. But we need to ensure we don’t spam the console with warnings about the same prop multiple times.

brad-decker commented:
@gaearon definitely. Will work on this shortly.

aweary commented:
@brad-decker any chance you could get the warning deduplicated? I closed #6255 so this PR is next in line if you want to continue working on it 😃

brad-decker commented:
@aweary Yep! will get this done today. Thanks for the nudge ;) :D

brad-decker commented:
@aweary added warning depulication, rebased and squashed down to two commits. First of which is the work from React-Europe 7 months ago then the new work in the new commit. Can squash and reword these to something that is more inline with what the React team would like for commit messages once reviewed.

brad-decker commented:
Any assistance with the failed tests and a point in the right direction for debugging my impact on them would be welcomed.

aweary commented:
Thanks @brad-decker! I'll come back to this later today to review again. I think there's a way we can avoid a couple of the `indexOf` checks to make this a bit faster.

>Any assistance with the failed tests and a point in the right direction for debugging my impact on them would be welcomed.

Go ahead and run the `./scripts/fiber/record-tests` script. We use it to track which unit tests are failing/passing for Fiber and if its not run when a new test is added CI will fail.

brad-decker commented:
@aweary got it. thanks

brad-decker commented:
@aweary, If this isn't planned to land soon thats cool. I'd like to still get your input and feedback on any improvements you think could be made to this when you have time to review. Thanks again!

aweary commented:
PropTypes have been removed from React core and now exist as a separate package `prop-types`. Any future feature requests, bug reports, or changes should be directed to the new repo at https://github.com/reactjs/prop-types.

Sorry about that @brad-decker!

aweary commented:
If you'd like to continue working on this I've opened a corresponding issue at https://github.com/reactjs/prop-types/issues/7
